### ISSUE TITLE
sample config.json

### ISSUE THREAD

**Issue by Just5KY** [qeeqbox/honeypots#18]
Can you add a sample config file that has all the options including ability to capture commands for some of the services you mentioned.

currently im using the modified [tpot.txt](https://github.com/qeeqbox/honeypots/blob/main/tpot.txt), what else can be added into this

```json
{
   "logs":"file,terminal,json",
   "logs_location":"/honeypots/logs",
   "filter":"",
   "interface":"",
   "honeypots":{
      "ftp":{
         "port":21,
         "ip":"0.0.0.0",
         "username":"ftp",
         "password":"anonymous",
         "log_file_name":"ftp.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "httpproxy":{
         "port":8080,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"admin",
         "log_file_name":"httpproxy.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "http":{
         "port":80,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"admin",
         "log_file_name":"http.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "https":{
         "port":443,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"admin",
         "log_file_name":"https.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "imap":{
         "port":143,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"imap.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "mysql":{
         "port":3306,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"mysql.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "pop3":{
         "port":110,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"pop3.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "postgres":{
         "port":5432,
         "ip":"0.0.0.0",
         "username":"postgres",
         "password":"123456",
         "log_file_name":"postgres.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "redis":{
         "port":6379,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"",
         "log_file_name":"redis.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "smb":{
         "port":445,
         "ip":"0.0.0.0",
         "username":"administrator",
         "password":"123456",
         "folders":"r&dbackup:/backup,secret:/secret,auth:/auth,vault:/vault",
         "log_file_name":"smb.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "smtp":{
         "port":25,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"smtp.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "socks5":{
         "port":1080,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"admin",
         "log_file_name":"socks5.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "ssh":{
         "port":22,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"ssh.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "telnet":{
         "port":23,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"telnet.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "vnc":{
         "port":5900,
         "ip":"0.0.0.0",
         "username":"administrator",
         "password":"123456",
         "log_file_name":"vnc.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "elastic":{
         "port":9200,
         "ip":"0.0.0.0",
         "username":"elastic",
         "password":"123456",
         "log_file_name":"elastic.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "mssql":{
         "port":1433,
         "ip":"0.0.0.0",
         "username":"sa",
         "password":"",
         "log_file_name":"mssql.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "ldap":{
         "port":389,
         "ip":"0.0.0.0",
         "username":"administrator",
         "password":"123456",
         "log_file_name":"ldap.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "ntp":{
         "port":123,
         "ip":"0.0.0.0",
         "username":"administrator",
         "password":"123456",
         "log_file_name":"ntp.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "memcache":{
         "port":11211,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"123456",
         "log_file_name":"memcache.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "oracle":{
         "port":1521,
         "ip":"0.0.0.0",
         "username":"bi",
         "password":"123456",
         "log_file_name":"oracle.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "snmp":{
         "port":161,
         "ip":"0.0.0.0",
         "username":"privUser",
         "password":"123456",
         "log_file_name":"snmp.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "irc":{
         "port":6667,
         "ip":"0.0.0.0",
         "username":"ircUser",
         "password":"123456",
         "log_file_name":"irc.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "sip":{
         "port":5060,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"123456",
         "log_file_name":"sip.log",
         "max_bytes":1000000,
         "backup_count":10
      },
      "sock5":{
         "port":1080,
         "ip":"0.0.0.0",
         "username":"admin",
         "password":"123456",
         "log_file_name":"sock5.log",
         "max_bytes":1000000,
         "backup_count":10
      },
   },
   "custom_filter":{
      "honeypots":{
         "change":{
            "server":"protocol"
         },
         "contains":[
            "protocol",
            "action",
            "src_ip",
            "src_port",
            "dest_ip",
            "dest_port"
         ],
         "remove":[
            
         ],
         "options":[
            "remove_errors",
            "remove_init",
            "remove_word_server"
         ]
      }
   }
}
```

**Comment by giga-a** [comment]
@Just5KY Yes! one second..

**Comment by giga-a** [comment]
Also, I added 2 more honeypots to that list (IRC and SIP)

**Comment by t3chn0m4g3** [comment]
@giga-a For some reason that went pass by me (so thanks to @Just5KY for opening this issue!). Just tested this out and noticed (although supported) it does not seem to work with SMTP... (even installed from PyPi to make sure my build was not faulty).

Logs from honeypots:
```
Creating network "honeypots_honeypots_local" with the default driver
Creating honeypots ... done
Attaching to honeypots
honeypots    | False
honeypots    | [x] Use [Enter] to exit or python3 -m honeypots --kill
honeypots    | [x] config.json file overrides --ip, --port, --username and --password
honeypots    | [x] QDNSServer, QFTPServer, QHTTPProxyServer, QHTTPServer, QHTTPSServer, QIMAPServer, QMysqlServer, QPOP3Server, QPostgresServer, QRedisServer, QSMBServer, QSMTPServer, QSOCKS5Server, QSSHServer, QTelnetServer, QVNCServer, QElasticServer, QMSSQLServer, QLDAPServer, QNTPServer, QMemcacheServer, QOracleServer, QSNMPServer, QSIPServer, QIRCServer running..
honeypots    | [x] Everything looks good!
honeypots    | {"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "172.20.254.127", "src_port": "45090", "timestamp": "2022-06-07T10:00:35.172121"}
```

Input from telnet session:
```
Trying 172.20.254.138...
Connected to 172.20.254.138.
Escape character is '^]'.
220 5e1f4a09dc14 Python SMTP proxy version 0.3
ehlo test
250-5e1f4a09dc14 Hello test
250-8BITMIME
250-AUTH LOGIN PLAIN
250 STARTTLS
mail from: john
250 OK
rcpt to: jane
250 OK
data
354 End data with <CR><LF>.<CR><LF>
subject: bla
test
.
250 OK
quit
221 Bye
Connection closed by foreign host.
```

Can you reproduce or is there something wrong on my end?

**Comment by Just5KY** [comment]
@t3chn0m4g3 yup i was able to reproduce this

```json
      "smtp":{
         "port":25,
         "ip":"0.0.0.0",
         "username":"root",
         "password":"123456",
         "log_file_name":"smtp.log",
         "max_bytes":1000000,
         "backup_count":10,
         "options":["capture_commands"]
      }
```

```bash
[root@sky ~]# telnet 152.67.188.122 25
Trying 152.67.188.122...
Connected to 152.67.188.122.
Escape character is '^]'.
220 55d52869acdf Python SMTP proxy version 0.3
ehlo test
250-55d52869acdf Hello test
250-8BITMIME
250-AUTH LOGIN PLAIN
250 STARTTLS
mail from: justsky
250 OK
rcpt to: github
250 OK
data
354 End data with <CR><LF>.<CR><LF>
subject: test
.
250 OK
quit
221 Bye
Connection closed by foreign host.
```

Only thing showing up is this
`{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "112.133.231.181", "src_port": "8670", "timestamp": "2022-06-07T11:01:27.932760"}`

**Comment by t3chn0m4g3** [comment]
@Just5KY Thanks for testing and reopening!

**Comment by Just5KY** [comment]
@t3chn0m4g3 did you test all the modules or just SMTP?

**Comment by t3chn0m4g3** [comment]
@Just5KY Just tested SMTP for now in the hopes it is just a config tweak before spending more time on this 😄 
@giga-a Happy to test more if it helps, though!

**Comment by Just5KY** [comment]
tried nmap

```bash
[root@sky ~]# nmap -A IP -p 25
Starting Nmap 7.70 ( https://nmap.org ) at 2022-06-07 11:12 UTC
Nmap scan report for IP
Host is up (0.070s latency).

PORT   STATE SERVICE    VERSION
25/tcp open  smtp-proxy Python SMTP Proxy 0.3
|_smtp-commands: 55d52869acdf Hello nmap.scanme.org,
|_smtp-ntlm-info: ERROR: Script execution failed (use -d to debug)
Warning: OSScan results may be unreliable because we could not find at least 1 o      pen and 1 closed port
Aggressive OS guesses: Linux 2.6.18 - 2.6.22 (94%), Linux 2.6.32 (91%), Linux 2.      6.18 (91%), Linux 2.6.32 or 3.10 (90%), Linux 2.6.39 (90%), Linux 3.10 - 3.12 (9      0%), Linux 3.4 (90%), Linux 3.7 (90%), Linux 4.4 (90%), Synology DiskStation Man      ager 5.1 (90%)
No exact OS matches for host (test conditions non-ideal).
Network Distance: 18 hops
```

logs
```json
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "IP", "src_port": "46914", "timestamp": "2022-06-07T11:12:27.665180"}
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "IP", "src_port": "46918", "timestamp": "2022-06-07T11:12:27.667040"}
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "IP", "src_port": "46922", "timestamp": "2022-06-07T11:12:27.667916"}
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "25", "protocol": "smtp", "src_ip": "IP", "src_port": "46920", "timestamp": "2022-06-07T11:12:27.668675"}
```

**Comment by t3chn0m4g3** [comment]
Pushed the changes (so far) to [T-Pot](https://github.com/telekom-security/tpotce/commit/a28dfec046205d6e6740382bb4d64923f2e58713), building and pushing the images just now.

**Comment by Just5KY** [comment]
checking other services using nmap se

**Comment by t3chn0m4g3** [comment]
@Just5KY According to the [Readme](https://github.com/qeeqbox/honeypots/blob/main/README.md) not all services do support logging `command_input`. My understanding is that https and snmp are not supported (supported list at the end).

**Comment by Just5KY** [comment]
My bad on snmp, i didnt forward UDP port in docker, and was scannning with `-sU` flag in Nmap, it works now

```bash
[root@sky ~]# nmap -A IP -p 161 -sU
Starting Nmap 7.70 ( https://nmap.org ) at 2022-06-07 12:18 UTC
Nmap scan report for IP
Host is up (0.065s latency).

PORT    STATE         SERVICE VERSION
161/udp open|filtered snmp
Too many fingerprints match this host to give specific OS details
```

```json
{"action": "query", "data": {"community": "public", "oids": "1.3.6.1.4.1.77.1.2.25", "version": "0"}, "dest_ip": "0.0.0.0", "dest_port": "161", "protocol": "snmp", "src_ip": "IP", "src_port": "29230", "status": "success", "timestamp": "2022-06-07T12:20:40.903073"}
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "161", "protocol": "snmp", "src_ip": "IP", "src_port": "29188", "status": "fail", "timestamp": "2022-06-07T12:20:40.905929"}
{"action": "query", "data": {"community": "public", "oids": "1.3.6.1.2.1.25.6.3.1", "version": "0"}, "dest_ip": "0.0.0.0", "dest_port": "161", "protocol": "snmp", "src_ip": "IP", "src_port": "29188", "status": "success", "timestamp": "2022-06-07T12:20:40.906499"}
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "161", "protocol": "snmp", "src_ip": "IP", "src_port": "29426", "status": "fail", "timestamp": "2022-06-07T12:20:45.908754"}
{"action": "query", "data": {"community": "public", "oids": "1.3.6.1.4.1.25506.2.12.1.1.1", "version": "0"}, "dest_ip": "0.0.0.0", "dest_port": "161", "protocol": "snmp", "src_ip": "IP", "src_port": "29426", "status": "success", "timestamp": "2022-06-07T12:20:45.909554"}
```

**Comment by giga-a** [comment]
[logs.txt](https://github.com/qeeqbox/honeypots/files/8855375/logs.txt)

**Comment by giga-a** [comment]
@t3chn0m4g3 & @Just5KY Alright! Done https://pypi.org/project/honeypots/0.52/ and I added the capture command logic to ssh, so now ssh honeypots will allow to let users in and then capturing the commands - It mimics ls, whoami, exit etc..

**Comment by giga-a** [comment]
@Just5KY Thanks for checking, I'm looking...

**Comment by t3chn0m4g3** [comment]
@Just5KY @giga-a I can confirm this for SSH, connection is logged, but then the connection hangs ...:
```
honeypots    | {"action": "login", "dest_ip": "0.0.0.0", "dest_port": "22", "password": "123456", "protocol": "ssh", "src_ip": "172.20.254.127", "src_port": "48212", "status": "success", "timestamp": "2022-06-08T11:34:10.780102", "username": "root"}
```

For FTP I see `success` for the login, the ftp client seems to receive a `failed`. But I think this is the same behavior as I encountered with my first tests a while back.

So far SMTP works great and as expected.

Thanks @giga-a :)

**Comment by t3chn0m4g3** [comment]
Within FTP the log differentiates between the actions for `command` and `connection`. Within SMTP this is not the case; as user I'd expect `connection` just for the connection and `command` for subsequent transmission within the established connection. This would also make it easier for filtering.

**Comment by giga-a** [comment]
@t3chn0m4g3 Thanks for letting me know! Are you using the Debian-based for running the honeypots and connecting to them? or only connecting to them?

**Comment by t3chn0m4g3** [comment]
@giga-a Mixed bag. Tested on same machine (passive mode, too) and from other machines. With Cyberduck on macOS I am getting a `550 Requested action not taken; file unavailable`:

```
honeypots    | {"action": "command", "data": {"args": "ftp", "cmd": "USER"}, "dest_ip": "0.0.0.0", "dest_port": "21", "protocol": "ftp", "src_ip": "172.20.254.195", "src_port": "63584", "timestamp": "2022-06-09T16:51:43.243533"}
honeypots    | {"action": "command", "data": {"args": "anonymous", "cmd": "PASS"}, "dest_ip": "0.0.0.0", "dest_port": "21", "protocol": "ftp", "src_ip": "172.20.254.195", "src_port": "63584", "timestamp": "2022-06-09T16:51:43.259025"}
honeypots    | {"action": "login", "dest_ip": "0.0.0.0", "dest_port": "21", "password": "anonymous", "protocol": "ftp", "src_ip": "172.20.254.195", "src_port": "63584", "status": "success", "timestamp": "2022-06-09T16:51:43.259230", "username": "ftp"}
```

I am running within Docker, though... let me check something...

**Comment by t3chn0m4g3** [comment]
... nevermind, even in `network_mode: "host"` the behavior does not change.

**Comment by t3chn0m4g3** [comment]
@giga-a To eliminate Docker as a possible culprit I tested against the PyPi install with the same results (ssh, ftp).

**Comment by giga-a** [comment]
@t3chn0m4g3 - Thanks for testing that 🙂 Just to recap, the ssh still hanging? And the ftp is giving you`550 Requested action not taken; file unavailable`?

**Comment by t3chn0m4g3** [comment]
@giga-a Yes, correct.

**Comment by giga-a** [comment]
@t3chn0m4g3 I think I figured the ssh one! If you do not pass --options capture_commands it will hang..

Have you tried this?

python3 -m honeypots --setup ssh --options capture_commands

**Comment by giga-a** [comment]
Okay, check 0.55 and let me know if works or not 

**ssh honeypot**

`capture_commands` now capture exec works with both ssh or scp

`scp -o HostKeyAlgorithms=+ssh-rsa test test@127.0.0.1:test`

`
{"action": "command", "data": {"command": "scp -t test"}, "dest_ip": "0.0.0.0", "dest_port": "22", "server": "ssh_server", "src_ip": "172.17.0.1", "src_port": "47538", "timestamp": "2022-06-10T05:09:36.402568"}
`

`ssh -o HostKeyAlgorithms=+ssh-rsa user@localhost ls`

`
{"action": "command", "data": {"command": "ls"}, "dest_ip": "0.0.0.0", "dest_port": "22", "server": "ssh_server", "src_ip": "172.17.0.1", "src_port": "47552", "timestamp": "2022-06-10T05:12:14.864117"}
`

`interactive` this one is experimental but makes it interactive, which heavily relays on how the user interacts and what operating system\client the user is using. So, it may fail at some point. This option is off unless you add `interactive` in the `options`

Finally, ssh authentication methods changed to password and public key; it logs the key fingerprint (It used to do password only). 

Tested on 
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye

**ftp honeypot**

Here is an example of logging in and using the PWD command

`
from ftplib import FTP
ftp = FTP('127.0.0.1')
ftp.login("admin","test")
print(ftp.pwd())
ftp.quit()
`

`
{"action": "connection", "dest_ip": "0.0.0.0", "dest_port": "21", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "timestamp": "2022-06-11T23:38:17.929400"}
{"action": "command", "data": {"args": "('admin',)", "cmd": "USER"}, "dest_ip": "0.0.0.0", "dest_port": "21", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "timestamp": "2022-06-11T23:38:17.929793"}
{"action": "command", "data": {"args": "('test',)", "cmd": "PASS"}, "dest_ip": "0.0.0.0", "dest_port": "21", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "timestamp": "2022-06-11T23:38:17.930990"}
{"action": "login", "dest_ip": "0.0.0.0", "dest_port": "21", "password": "test", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "status": "failed", "timestamp": "2022-06-11T23:38:17.931084", "username": "admin"}
{"action": "command", "data": {"args": "()", "cmd": "PWD"}, "dest_ip": "0.0.0.0", "dest_port": "21", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "timestamp": "2022-06-11T23:38:17.931444"}
{"action": "command", "data": {"args": "()", "cmd": "QUIT"}, "dest_ip": "0.0.0.0", "dest_port": "21", "server": "ftp_server", "src_ip": "172.17.0.1", "src_port": "35472", "timestamp": "2022-06-11T23:38:17.932002"}
`

Tested on 
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye

**Comment by giga-a** [comment]
@t3chn0m4g3 That's great!! Yes, the prompt is cool :smiley: I'm currently making a different PyPI package that handles it. The break was caused by the timeout I placed, and I will remove timeouts in the next update. 

> FTP: Connection now works incl. commands, however for some clients the passive mode needs to be used, which is expected.

Okay, I will check the ftp

> Which services currently support the interactive option?

Currently, ssh, but I'm going to add telnet that way it emulates some of the IoT devices.

I will be back..
